rust/src/grid/mod.rs

```diff
@@ -0,0 +1,142 @@
+pub mod builder;
+mod debug;
+mod freq;
+mod mutate;
+
+use crate::cell::Cell;
+use crate::piece::Piece;
+use crate::quota::Quota;
+
+pub struct Grid<'a> {
+    pieces: Vec<Piece>,
+    qcol: Vec<Quota>,
+    qrow: Vec<Quota>,
+    row_pieces: Vec<Vec<(u8, usize)>>,
+    cells: Vec<Vec<Cell>>,
+    groups: &'a Vec<Vec<u8>>,
+}
+
+impl<'a> Grid<'a> {
+    pub fn new(
+        cols: &Vec<i32>,
+        rows: &Vec<i32>,
+        groups: &'a Vec<Vec<u8>>,
+    ) -> Self {
+        let size = cols.len();
+        let qcol = cols.iter().map(|v| Quota::new(*v, size)).collect();
+        let qrow = rows.iter().map(|v| Quota::new(*v, size)).collect();
+
+        let pieces = builder::pieces(size, groups);
+        let row_pieces = builder::row_pieces(groups);
+        let cells = vec![vec![Cell::None; size]; size];
+
+        Self { pieces, qcol, qrow, row_pieces, cells, groups }
+    }
+
+    fn size(&self) -> usize {
+        self.qcol.len()
+    }
+
+    // execute one row-forcing move
+    // returns true if a move is executed
+    pub fn row_forcing_move(&mut self) -> bool {
```

Comment:
the core idea of this implementation: only make forced moves, and hence the board state is always known to be correct and on the path to the answer. Constraints are stated in README